python/ql/src/experimental/Security/CWE-295/UrlopenWithoutValidation.ql

```diff
@@ -0,0 +1,29 @@
+/**
+ * @name Attempt to open url without certificate validation
+ * @description Opening a url without certificate validation can allow man-in-the-middle attacks.
+ * @kind problem
+ * @problem.severity error
+ * @id py/urlopen-without-cert-validation
+ * @tags security
+ *       external/cwe/cwe-295
+ */
+
+import python
+
+from
+  FunctionValue urlopen, CallNode call, ControlFlowNode http_arg, StringValue http_string,
+  Value empty
+where
+  urlopen = Value::named(["urllib2.request.urlopen", "urllib.request.urlopen"]) and
+  http_arg = urlopen.getArgumentForCall(call, 0) and
+  http_arg.pointsTo(http_string) and
+  http_string.getText().matches("https://%") and
+  (
+    not exists(Value verify |
+      verify = call.getArgByName(["cafile", "capath", "context"]).pointsTo()
+    )
+    or
+    empty = call.getArgByName(["cafile", "capath", "context"]).pointsTo() and
+    empty = Value::none_()
+  )
```

Comment:
Our current approach to modelling libraries is to use [API graphs](https://codeql.github.com/docs/codeql-language-guides/using-api-graphs-in-python/). 

In this case, I think it should be fairly simple to rewrite the code to use API graphs. The calls you're interested in can be found using 
```ql
API::moduleImport(["urllib","urllib2"]).getMember("request").getMember("urlopen").getACall()
```
For the value that is being passed to the named argument, I would either see if the named argument is absent altogether (in which case the default is used), or if it is present and equal to `API::builtin("None").getAUse()`. That should mostly replicate the behaviour you have above.

Reply:
Thanks for the advice, API graphs are a really cool feature! I added the changes you recommended, and reformatted the code a bit (made sure to remember the autoformatting too 👍 ).

I weirdly did not catch one of my test cases when I used `API::builtin("None").getAUse()` insead of comparing the value of the argument to `Value::none_()` however. I kept the latter method, which worked, and commented out the former. If you have any ideas why that might be the case that would be super helpful, but otherwise the method I am currently using works anyway (I will remove the comments once I get feedback on this decision). Thanks again, and let me know if there is anything else I need to fix!

Reply:
Ah, it's possible we have a bug in `API::builtin` where we fail to include things like `None` (since `API::builtin` is really a model of the `builtins` module, but `None` is kind of special). I'll see if I can make a quick fix for that.

Reply:
#5639 should fix the problem with `None`.

Incidentally, for this one it might be better to use `getAnImmediateUse` rather than `getAUse`. The difference here is that the latter will also take some (possibly unsound) dataflow into account, and this may lead to undesirable false negatives.

Reply:
@tausbn new pull request is here: #5644. I used `getAnImmediateUse` instead and modified the test slightly to allow for this. I also kept `API::builtin("None").getAUse()` since you have opened a PR to fix this. 👍